rclcpp/src/rclcpp/time_source.cpp

```diff
@@ -192,6 +192,7 @@ void TimeSource::set_clock(
     enable_ros_time(clock);
   }
 
+  std::lock_guard<std::mutex> clock_guard(clock->get_clock_mutex());
```

Comment:
One problem here is that the clock could, in theory, be changed between the call to `{enable,disable}_ros_time` above, and `rcl_set_ros_time_override`.

What I'll suggest is something a little different.  Create two new `private` methods to this class called `disable_ros_time_nolock()` and `enable_ros_time_nolock()`, which has the current contents of the respective methods.  Then in `disable_ros_time()`, take the lock and call `disable_ros_time_nolock()`; do the same for `enable_ros_time()`.  Finally, change `set_clock()` to take the lock right at the beginning, and then call `disable_ros_time_nolock()` or `enable_ros_time_nolock()` as appropriate.

Reply:
`disable_ros_time(clock)` and `enable_ros_time(clock)` are only used in `set_clock`.
So, how about merging them into `set_clock` to make things clear.